File: 001201-001400/001247.交换字符使得字符串相同/1247.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {number}
 */
// 方法一：用时 64 ms
// 执行用时超过了 85 %
// 消耗内存超过了 16 %
// 难度: 中等
var minimumSwap = function(s1, s2) {
  if (s1.length != s2.length) {
    return -1
  }
  let xy = 0
  let yx = 0
  for (let i = 0, j = 0; i < s1.length && j < s2.length; i ++, j ++) {
    if (s1[i] == 'x' && s2[j] == 'y') {
      xy ++
    } else if (s1[i] == 'y' && s2[j] == 'x') {
      yx ++
    }
  }
  if ((xy + yx) % 2 == 1) {
    return -1
  }
  let ret = Math.floor(xy / 2) + Math.floor(yx / 2)
  if (xy % 2 == 1) {
    ret += 2
  }
  return ret
};

console.log(minimumSwap('xx', 'yy'))
console.log(minimumSwap('xy', 'yx'))
console.log(minimumSwap('xx', 'xy'))
console.log(minimumSwap('xxyyxyxyxx', 'xyyxyxxxyx'))

File: 001201-001400/001247.交换字符使得字符串相同/1247.go
package main

import "fmt"

// 方法一：用时 0 ms
// 执行用时超过了 100 %
// 消耗内存超过了 60 %
// 难度: 中等
func minimumSwap(s1 string, s2 string) int {
	if len(s1) != len(s2) {
		return -1
	}
	xy := 0
	yx := 0
	for i, j := 0, 0; i < len(s1) && j < len(s2); i ++ {
		if s1[i] == 'x' && s2[j] == 'y' {
			xy ++
		} else if s1[i] == 'y' && s2[j] == 'x' {
			yx ++
		}
		j ++
	}
	if (xy + yx) % 2 != 0 {
		return -1
	}
	ret := xy / 2 + yx / 2
	if xy % 2 != 0 {
		ret += 2
	}
	return ret
}

func main ()  {
	fmt.Println(minimumSwap("xx", "yy"))
	fmt.Println(minimumSwap("xy", "yx"))
	fmt.Println(minimumSwap("xy", "xx"))
	fmt.Println(minimumSwap("xxyyxyxyxx", "xyyxyxxxyx"))
}